test/integration/ws/common.dart

```diff
@@ -225,61 +227,169 @@ void runCommonWebSocketIntegrationTests({connect(Uri uri), int port}) {
-    socket.add(_closeRequest(4001, 'Closed by server.'));
-    await socket.done;
-    await socket.close(4002, 'Late close.');
-    expect(socket.closeCode, equals(4001));
-    expect(socket.closeReason, equals('Closed by server.'));
+    var webSocket = await connect(closeUri);
+    webSocket.add(_closeRequest(4001, 'Closed by server.'));
+    await webSocket.done;
+    await webSocket.close(4002, 'Late close.');
+    expect(webSocket.closeCode, equals(4001));
+    expect(webSocket.closeReason, equals('Closed by server.'));
   });
 
   test('should ignore close() calls after the first', () async {
-    WSocket socket = await connect(echoUri);
-    await socket.close(4001, 'Custom close.');
-    await socket.close(4002, 'Late close.');
-    expect(socket.closeCode, equals(4001));
-    expect(socket.closeReason, equals('Custom close.'));
+    var webSocket = await connect(echoUri);
+    await webSocket.close(4001, 'Custom close.');
+    await webSocket.close(4002, 'Late close.');
+    expect(webSocket.closeCode, equals(4001));
+    expect(webSocket.closeReason, equals('Custom close.'));
   });
 
   test(
       'should report the close code and reason that the server used when closing the connection',
       () async {
-    WSocket socket = await connect(closeUri);
+    var socket = await connect(closeUri);
     socket.add(_closeRequest(4001, 'Closed by server.'));
     await socket.done;
     expect(socket.closeCode, equals(4001));
     expect(socket.closeReason, equals('Closed by server.'));
   });
+
+  test('message events should be discarded prior to a subscription', () async {
+    var webSocket = await connect(echoUri);
+
+    webSocket.add('1');
+    webSocket.add('2');
+    await new Future.delayed(new Duration(milliseconds: 500));
+
+    var messages = [];
+    webSocket.listen((data) {
+      messages.add(data);
+    });
+
+    webSocket.add('3');
+    webSocket.add('4');
+    await new Future.delayed(new Duration(milliseconds: 500));
+
+    await webSocket.close();
+    expect(messages, orderedEquals(['3', '4']));
+  });
+
+  test(
+      'the first event should be received if a subscription is made immediately',
+      () async {
+    var webSocket = await connect(echoUri);
+
+    var c = new Completer();
+    webSocket.listen((data) {
+      c.complete(data);
+    });
+    webSocket.add('first');
+
+    expect(await c.future, equals('first'));
+    await webSocket.close();
+  });
+
+  test('all event streams should respect pause() and resume() signals',
+      () async {
+    var webSocket = await connect(echoUri);
+    var messages = [];
+
+    // no subscription yet, messages should be discarded
+    webSocket.add('1');
+    await new Future.delayed(new Duration(milliseconds: 500));
+
+    // setup a subscription, messages should be recorded
+    var sub = webSocket.listen((data) {
+      messages.add(data);
+    });
+    webSocket.add('2');
+    await new Future.delayed(new Duration(milliseconds: 500));
```

Comment:
agreed, I'm going to try to get these down. It was necessary to get the SockJS suite passing, but that was before I was properly cleaning up all socket connections, so it might be better now